### PR TITLE
Rtl-improvement

### PR DESCRIPTION
**Pull request by wissog**
some coding improvements

### CODE REVIEW

File: src/redux/modules/tags.js
Comment (davgur) [inline src/redux/modules/tags.js:1]: this is major change of this PR (same for source). I remove building of  _getByID_ , _getPathByID_  and  _getPath_ because its broke hydration of SSR html
@wissog @bbfsdev

Comment (wissog) [reply src/redux/modules/tags.js:1]: ok, so what is needed to merge it?
here are some improvements which seem not directly related to what you are doing

File: src/components/Sections/Events/Filters.js [hunk 1]
```diff
@@ -15,12 +14,14 @@ import OriginalLanguageFilter from '../../FiltersAside/OriginalLanguageFilter/Or
...
 import ContentTypesFilter from './ContentTypesFilter';
 
-const Filters = ({ namespace, baseParams, t }) => {
+const Filters = ({ namespace, baseParams }) => {
   const [isHydrated, setIsHydrated] = useState(false);
 
+  const { t }        = useTranslation();
   const isReady      = useSelector(state => selectors.isReady(state.filtersAside, namespace));
   const { wip, err } = useSelector(state => selectors.getWipErr(state.filtersAside, namespace));
-  const selected     = useSelector(state => filters.getNotEmptyFilters(state.filters, namespace), isEqual);
```
Comment (bbfsdev) [inline src/components/Sections/Events/Filters.js:23]: Why did you remove here isEqual?

File: src/components/Sections/Events/Filters.js [hunk 2]
```diff
@@ -31,14 +32,16 @@ const Filters = ({ namespace, baseParams, t }) => {
     }
   }, [dispatch, isReady]);
 
+  const selLen = selected.reduce((acc, x) => acc + x.values.length, 0);
   useEffect(() => {
-    if (isHydrated && isReady) {
+    if (isHydrated && isReady && prevSelRef.current !== selLen) {
```
Comment (bbfsdev) [inline src/components/Sections/Events/Filters.js:37]: What is the fix here that did not work earlier?

File: src/components/Sections/Lesson/Filters.js [hunk 3]
```diff
@@ -27,11 +29,13 @@ const Filters = ({ namespace, baseParams, t }) => {
     }
   }, [dispatch, isReady, wip, err]);
 
+  const selLen = selected.reduce((acc, x) => acc + x.values.length, 0);
   useEffect(() => {
-    if (isHydrated && isReady) {
+    if (isHydrated && isReady && prevSelRef.current !== selLen) {
```
Comment (bbfsdev) [inline src/components/Sections/Lesson/Filters.js:34]: What are you checking with this?
What is prevSelRef?

File: src/components/Sections/Lessons/Filters.js [hunk 4]
```diff
@@ -46,8 +48,9 @@ const Filters = ({ namespace, baseParams, t }) => {
     }
   }, [isReady, baseParams, wip, err, namespace, dispatch]);
 
+  const selLen = selected.reduce((acc, x) => acc + x.values.length, 0);
   useEffect(() => {
-    if (isHydrated && isReady) {
+    if (isHydrated && isReady && prevSelRef.current !== selLen) {
```
Comment (bbfsdev) [inline src/components/Sections/Lessons/Filters.js:53]: Copy paste of the previous code.... why keep duplicates?

File: src/helpers/utils.js [hunk 5]
```diff
@@ -461,3 +463,20 @@ export const getSourcesCollections = (sources, getPathById) =>
       return acc;
     }, {})).filter(source => source && source.children && source.children.length);
 
+// helper function to filter array and get unique values - arr.filter(distinct)
+export const distinct = (value, index, arr) => arr.indexOf(value) === index;
```
Comment (bbfsdev) [inline src/helpers/utils.js:467]: Don't use this this is not efficient, see comment above.

File: src/helpers/utils.js [hunk 6]
```diff
@@ -461,3 +463,20 @@ export const getSourcesCollections = (sources, getPathById) =>
...
+
+export const renderTags = tagNames => (
+  tagNames?.length > 0 &&
+  <div className="unit-tags-bar">
+    {
+      tagNames
```
Comment (bbfsdev) [inline src/helpers/utils.js:473]: [...new Set(tagNames)].map((tag, index) => .....

File: src/components/Sections/Topics/SubTopics.js [hunk 7]
```diff
@@ -29,35 +29,28 @@ const getItemsRecursive = (rootID, getById, base) => {
...
   const selected        = useMemo(() => selectedFilters?.values || [], [selectedFilters]);
 
-  console.log('SELECTED', selected);
-
   const root  = getTagById(rootID);
   const items = useMemo(() => getItemsRecursive(rootID, getTagById, baseItems) || [], [rootID, getTagById, baseItems]);
 
   useEffect(() => setQuery(null), []);
-
-  useEffect(() => {
```
Comment (bbfsdev) [inline src/components/Sections/Topics/SubTopics.js:51]: Why did you remove this effect?

File: src/components/Sections/Topics/RenderPage.js [hunk 8]
```diff
@@ -27,7 +27,7 @@ const RenderPage = ({ t }) => {
   const breadCrumbSections = [{ id: '', label: t('nav.sidebar.topics') }, ...tagPath].map(getBreadCrumbSection);
 
   const breadCrumbIcon = `${isLanguageRtl(language) ? 'left' : 'right'} angle`;
-
+  const baseParams     = useMemo(() => ({ tag: id }), [id]);
```
Comment (bbfsdev) [inline src/components/Sections/Topics/RenderPage.js:30]: Why useMemo here?

File: src/components/Sections/Sketches/Filters.js [hunk 9]
```diff
@@ -32,14 +34,16 @@ const Filters = ({ namespace, baseParams, t }) => {
     }
   }, [namespace, _isReady, baseParams, dispatch]);
 
+  const selLen = selected.reduce((acc, x) => acc + x.values.length, 0);
   useEffect(() => {
-    if (isHydrated && isReady) {
+    if (isHydrated && isReady && prevSelRef.current !== selLen) {
```
Comment (bbfsdev) [inline src/components/Sections/Sketches/Filters.js:39]: Again, not sure what that does, can this be refactored?

File: src/components/Pages/WithPlayer/widgets/UnitMaterials/Sketches.js [hunk 10]
```diff
@@ -51,7 +51,7 @@ class Sketches extends React.Component {
   static getFilesLanguages = (zipFiles, contentLanguage, uiLanguage) => {
     const languages = zipFiles
       .map(file => file.language)
-      .filter((v, i, a) => a.indexOf(v) === i);
+      .filter(distinct);
```
[the remark this answers is not in the document]
Comment (wissog) [reply src/components/Pages/WithPlayer/widgets/UnitMaterials/Sketches.js:54]: 1k of sketches per file???? we are far away from there...

[the remark this answers is not in the document]
Comment (davgur) [reply src/components/Pages/WithPlayer/widgets/UnitMaterials/Sketches.js]: done

File: src/components/Sections/Events/Filters.js (hunk 1 above)
Comment (davgur) [reply src/components/Sections/Events/Filters.js:23]: not need this additional checking, I use Ref for compare with preview state

File: src/components/Sections/Events/Filters.js (hunk 2 above)
Comment (davgur) [reply src/components/Sections/Events/Filters.js:37]: from one side I need fetch data when filters was change (add/remove checkbox) from other side I don't need if nothing was select (thats was happened before). For don't push to trigger dependency complicate object I use length of filters values for compare

File: src/components/Sections/Lessons/Filters.js (hunk 4 above)
Comment (davgur) [reply src/components/Sections/Lessons/Filters.js:53]: not clear duplicate with what?

File: src/helpers/utils.js (hunk 5 above)
Comment (davgur) [reply src/helpers/utils.js]: removed

File: src/components/Sections/Topics/SubTopics.js (hunk 7 above)
Comment (davgur) [reply src/components/Sections/Topics/SubTopics.js:51]: its not depend on third side so not need effect

File: src/components/Sections/Topics/RenderPage.js (hunk 8 above)
Comment (davgur) [reply src/components/Sections/Topics/RenderPage.js:30]: we push it to props of child node, I think its can make additional render of child

File: src/components/Sections/Publications/tabs/Articles/Unit.js [hunk 11]
```diff
@@ -17,27 +17,14 @@ import WipErr from '../../../../shared/WipErr/WipErr';
...
+import { renderTags } from '../../../../../helpers/utils';
 import { selectors as tagSelectors } from '../../../../../redux/modules/tags';
-import Link from '../../../../Language/MultiLanguageLink';
 
 const renderHeader = (unit, tagNames, t, language) => {
   const isRtl    = isLanguageRtl(language);
   const position = isRtl ? 'right' : 'left';
   const subText2 = t(`publications.header.subtext2`);
 
-  const renderTags = () => (
```
[the remark this answers is not in the document]
Comment (davgur) [reply src/components/Sections/Publications/tabs/Articles/Unit.js:28]: make special component for this

**Review by davgur — APPROVED** [review]

File: src/components/Sections/Events/Filters.js (hunk 2 above)
Comment (bbfsdev) [reply src/components/Sections/Events/Filters.js:37]: What is someone adds one checkbox and removes another, that way the length will be the same and you will not update anything?

File: src/helpers/utils.js (hunk 5 above)
Comment (bbfsdev) [reply src/helpers/utils.js]: I see the old code here, I don't see you have removed it. Did you update the PR?

File: src/components/Sections/Events/Filters.js (hunk 2 above)
Comment (davgur) [reply src/components/Sections/Events/Filters.js:37]: it will 2 different updates

File: src/helpers/utils.js (hunk 6 above)
Comment (davgur) [reply src/helpers/utils.js]: we can take tags from unit and from Label

File: src/helpers/utils.js (hunk 5 above)
Comment (davgur) [reply src/helpers/utils.js]: https://github.com/Bnei-Baruch/kmedia-mdb/pull/985/files#diff-6f75c83f4834bc3c6d394b362d541df2e842ff1ab52b1112a2a3aae9541498b9

Comment (davgur) [reply src/helpers/utils.js]: [https://github.com/Bnei-Baruch/kmedia-mdb/pull/985/files#diff-6f75c83f4834bc3c6d394b362d541df2e842ff1ab52b1112a2a3aae9541498b9](changes of utils)

**Review by bbfsdev — APPROVED** [review]
